File: packages/python/tryaii_dre/scoring/priorities.py
"""
User priority system for model selection.

Priorities let users express what matters to them (quality, cost, speed)
on a 1-5 scale. These get transformed into weights that influence scoring.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Priorities:
    """
    User priorities for model selection.

    Each value is on a 1-5 scale:
        1 = don't care about this dimension
        3 = balanced (default)
        5 = this is critical

    Examples:
        Priorities(quality=5, cost=1, speed=1)  # Best quality, ignore cost/speed
        Priorities(quality=2, cost=5, speed=4)  # Budget-focused, prefer fast
        Priorities(quality=3, cost=3, speed=3)  # Balanced (default)
    """

    quality: int = 3
    cost: int = 3
    speed: int = 3

    def __post_init__(self):
        for field_name in ("quality", "cost", "speed"):
            value = getattr(self, field_name)
            if not isinstance(value, (int, float)):
                raise TypeError(f"{field_name} must be a number, got {type(value)}")
            clamped = max(1, min(5, int(value)))
            object.__setattr__(self, field_name, clamped)

    @property
    def quality_weight(self) -> float:
        """Quality weight: 0.3 to 1.2 (always has baseline influence)."""
        return 0.3 + (self.quality / 5) * 0.9

    @property
    def cost_weight(self) -> float:
        """Cost weight: 0.1 to 1.0 (can be fully suppressed)."""
        return 0.1 + (self.cost / 5) * 0.9

    @property
    def speed_weight(self) -> float:
        """Speed weight: 0.1 to 1.0 (can be fully suppressed)."""
        return 0.1 + (self.speed / 5) * 0.9

    def to_dict(self) -> dict[str, int]:
        return {"quality": self.quality, "cost": self.cost, "speed": self.speed}

    @classmethod
    def from_dict(cls, d: dict) -> Priorities:
        return cls(
            quality=d.get("quality", 3),
            cost=d.get("cost", 3),
            speed=d.get("speed", 3),
        )

    @classmethod
    def performance(cls) -> Priorities:
        """Preset: maximize quality, ignore cost and speed."""
        return cls(quality=5, cost=1, speed=1)

    @classmethod
    def budget(cls) -> Priorities:
        """Preset: minimize cost, moderate quality."""
        return cls(quality=2, cost=5, speed=3)

    @classmethod
    def fast(cls) -> Priorities:
        """Preset: fastest response, moderate quality."""
        return cls(quality=2, cost=3, speed=5)

    @classmethod
    def balanced(cls) -> Priorities:
        """Preset: balanced across all dimensions."""
        return cls(quality=3, cost=3, speed=3)


DEFAULT_PRIORITIES = Priorities(quality=3, cost=3, speed=3)

File: packages/node/src/scoring/priorities.ts
/**
 * User priority system for model selection.
 *
 * Priorities let users express what matters to them (quality, cost, speed)
 * on a 1-5 scale. These get transformed into weights that influence scoring.
 */

export interface PrioritiesData {
  quality: number;
  cost: number;
  speed: number;
}

export class Priorities {
  /**
   * Each value is on a 1-5 scale:
   *   1 = don't care about this dimension
   *   3 = balanced (default)
   *   5 = this is critical
   */
  readonly quality: number;
  readonly cost: number;
  readonly speed: number;

  constructor(quality = 3, cost = 3, speed = 3) {
    this.quality = Math.max(1, Math.min(5, Math.round(quality)));
    this.cost = Math.max(1, Math.min(5, Math.round(cost)));
    this.speed = Math.max(1, Math.min(5, Math.round(speed)));
  }

  /** Quality weight: 0.3 to 1.2 (always has baseline influence). */
  get qualityWeight(): number {
    return 0.3 + (this.quality / 5) * 0.9;
  }

  /** Cost weight: 0.1 to 1.0 (can be fully suppressed). */
  get costWeight(): number {
    return 0.1 + (this.cost / 5) * 0.9;
  }

  /** Speed weight: 0.1 to 1.0 (can be fully suppressed). */
  get speedWeight(): number {
    return 0.1 + (this.speed / 5) * 0.9;
  }

  toDict(): PrioritiesData {
    return { quality: this.quality, cost: this.cost, speed: this.speed };
  }

  static fromDict(d: Partial<PrioritiesData>): Priorities {
    return new Priorities(d.quality ?? 3, d.cost ?? 3, d.speed ?? 3);
  }

  /** Preset: maximize quality, ignore cost and speed. */
  static performance(): Priorities {
    return new Priorities(5, 1, 1);
  }

  /** Preset: minimize cost, moderate quality. */
  static budget(): Priorities {
    return new Priorities(2, 5, 3);
  }

  /** Preset: fastest response, moderate quality. */
  static fast(): Priorities {
    return new Priorities(2, 3, 5);
  }

  /** Preset: balanced across all dimensions. */
  static balanced(): Priorities {
    return new Priorities(3, 3, 3);
  }
}

export const DEFAULT_PRIORITIES = new Priorities(3, 3, 3);
